### ISSUE TITLE
UKCP download from CEDA - transition to OPenDAP?

### ISSUE THREAD

**Issue by j08lue** [os-climate/hazard#133]
The Hazard UKCP source data retrieval is currently relying on CEDA's legacy FTP service.

I remember that there was some reason for preferring that over OPenDAP, no, @ciaransweet?

CEDA is now asking whether we can transition - what should we tell them?

> You are a registered user with the Centre for Environmental Data Analysis ([CEDA](https://archive.ceda.ac.uk/)) and we believe that you have used our FTP service ([ftp.ceda.ac.uk](http://ftp.ceda.ac.uk/)) to download data in the past.
> 
> We are seeking input from our FTP users to decide if the service is still required. It is currently unavailable due to a security incident. Please review the alternative ways of downloading data below:
> 
> You can use our [main download service (CEDA OPenDAP)](https://help.ceda.ac.uk/article/99-download-data-from-ceda-archives). The most suitable alternative to FTP for bulk downloads is to use the [wget command](https://help.ceda.ac.uk/article/4431-ceda-archive-web-download-and-services). 
> 
> If you are a [JASMIN](https://jasmin.ac.uk/) user, you can access a large range of CEDA data directly - you can either use that data on JASMIN machines or can download it using a variety of methods. [Look at the JASMIN access eligibility criteria](https://jasmin.ac.uk/users/access/) if you are not an existing user but would like to explore this option.
> 
> If these alternative methods are not suitable for you, please get in touch with the following information:
> Why is FTP more suitable for your needs?
> Are there any other download methods (such as rsync) you would like us to offer?
> Please send us your feedback as soon as possible. Many thanks.
> 
> CEDA user support
> [support@ceda.ac.uk](mailto:support@ceda.ac.uk)

**Comment by ciaransweet** [comment]
Yep already responded to them around this. The reason why OPenDAP isn't really suitable is you need to know exactly which file it is you want to download. AFAIK there is no search functionality.

My response:

> Hi
> 
> Yes I am still using the FTP service
> 
> This is core to our workflows as OPenDAP does not support exploring what data is available, it still requires you to know exactly what data you want before downloading it and that isn't possible programmatically. (As far as I know, happy to be shown otherwise!)
> 
> Right now, we download UKCP18 data files based on parameters such as the resolution, variable, and most importantly year the data maps to. Unfortunately UKCP18 data is saved in chunks that contain year overlaps. For example, if I wanted say all of 2069's data, you can see it's contained in two files:
> 
> 
![image](https://github.com/user-attachments/assets/529cca79-5c26-4a55-a8e8-ba81ebf499ab)
> 
> OPenDAP has no way of figuring this out, but I can do this by listing files via an FTP library, parsing the filenames, and determining that those 2 files contain data for 2069.
> 
> I think for any programmatic use case dependent on on-the-fly access to CEDA data, FTP is crucial, till there is a better way to query the data holdings without manual intervention.
> 
> I hope the FTP gets better soon! Happy to have a call etc to discuss this!
> 
> Cheers

**Comment by j08lue** [comment]
Looks like they recommend to get the opeNdap resource URLs from their CSW catalog https://archive.ceda.ac.uk/tools/ 👉 https://csw.ceda.ac.uk/geonetwork/srv/eng/csw?SERVICE=CSW&VERSION=2.0.2&REQUEST=GetCapabilities

**Comment by j08lue** [comment]
@ciaransweet found that you can get a "directory" listing with `?json` https://data.ceda.ac.uk/badc/ukcp18/data/land-gcm/uk/60km/rcp85?json

Phew

**Comment by ciaransweet** [comment]
Currently waiting in a support thread/queue to resolve how to get an access token to the CEDA Archives for downloading, FTP is totally removed now and the directory listing method only solves half our problem.
